kroxylicious/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyFrontendHandler.java

```diff
@@ -21,40 +40,126 @@
 import io.netty.channel.ChannelInboundHandlerAdapter;
 import io.netty.channel.ChannelOption;
 import io.netty.channel.ChannelPipeline;
+import io.netty.handler.codec.haproxy.HAProxyMessage;
 import io.netty.handler.logging.LoggingHandler;
+import io.netty.handler.ssl.SniCompletionEvent;
 
-public class KafkaProxyFrontendHandler extends ChannelInboundHandlerAdapter {
+public class KafkaProxyFrontendHandler
+        extends ChannelInboundHandlerAdapter
+        implements NetFilter.NetFilterContext {
 
     private static final Logger LOGGER = LoggerFactory.getLogger(KafkaProxyFrontendHandler.class);
 
-    private final String remoteHost;
-    private final int remotePort;
-    private final CorrelationManager correlationManager;
+    /** Cache ApiVersions response which we use when returning ApiVersions ourselves */
+    private static final ApiVersionsResponseData API_VERSIONS_RESPONSE;
```

Comment:
Yeah, what I have implemented is the simplest thing that could possibly work (because this PR is already pretty big). I think the best solution would be for the proxy to maintain its own persistent connection to each cluster on which it makes an ApiVersions request that is then cached for the duration of that connection. This is complicated but it alleviates the need for configuring this and lets the proxy behave better when brokers get upgraded.

It's actually a bit more complicated still if we want to support [Kafka's feature flags](https://cwiki.apache.org/confluence/display/KAFKA/KIP-584%3A+Versioning+scheme+for+features) properly, because those can change without a broker restart, but something like this would, for sure, be better than what I've implemented here. I will create an issue for this as/when this PR gets merged.